homeassistant/components/keba/__init__.py

```diff
@@ -0,0 +1,229 @@
+    except ValueError as ex:
+        _LOGGER.warning("Could not set failsafe mode %s", ex)
+
+    # Register services to hass
+    async def execute_service(call):
+        """Execute a service to KEBA charging station.
+
+        This must be a member function as we need access to the keba
+        object here.
+        """
+        function_name = _SERVICE_MAP[call.service]
+        function_call = getattr(keba, function_name)
+        await function_call(call.data)
+
+    for service in _SERVICE_MAP:
+        hass.services.async_register(DOMAIN, service, execute_service)
+
+    # Load components
+    for domain in SUPPORTED_COMPONENTS:
+        hass.async_create_task(
+            discovery.async_load_platform(hass, domain, DOMAIN, {}, config)
+        )
+
+    # Start periodic polling of charging station data
+    keba.start_periodic_request()
+
+    return True
+
+
+class KebaHandler(KebaKeContact):
+    """Representation of a KEBA charging station connection."""
+
+    def __init__(self, hass, host, rfid, refresh_interval):
+        """Constructor."""
+        super().__init__(host, self.hass_callback)
+
+        self._update_listeners = []
+        self._hass = hass
+        self.rfid = rfid
+        self.device_name = "keba_wallbox_"
+
+        # Ensure at least MAX_POLLING_INTERVAL seconds delay
+        self._refresh_interval = max(MAX_POLLING_INTERVAL, refresh_interval)
+        self._fast_polling_count = MAX_FAST_POLLING_COUNT
+        self._polling_task = None
+
+    def start_periodic_request(self):
+        """Start periodic data polling."""
+        self._polling_task = self._hass.loop.create_task(self._periodic_request())
+
+    async def _periodic_request(self):
+        """Send  periodic update requests."""
+        await self.request_data()
+
+        if self._fast_polling_count < MAX_FAST_POLLING_COUNT:
+            self._fast_polling_count += 1
+            _LOGGER.debug("Periodic data request executed, now wait for " "2 seconds.")
+            await asyncio.sleep(2)
+        else:
+            _LOGGER.debug(
+                "Periodic data request executed, now wait for " "%s seconds.",
+                str(self._refresh_interval),
+            )
+            await asyncio.sleep(self._refresh_interval)
+
+        _LOGGER.debug("Periodic data request rescheduled.")
+        self._polling_task = self._hass.loop.create_task(self._periodic_request())
+
+    async def setup(self, loop=None):
+        """Initialize KebaHandler object."""
+        await super().setup(loop)
+
+        # Request initial values and extract serial number
+        await self.request_data()
+        if self.get_value("Serial") is not None:
+            self.device_name = f"keba_wallbox_{self.get_value('Serial')}"
+        else:
+            _LOGGER.warning(
+                "Could not load the serial number of the "
+                "charging station, unique id will be wrong"
```

Comment:
Do we set `unique_id` to `None` in this case?

Reply:
no it should be "keba_wallbox_" see line 113. Or do you think it is better to stop the process of creating all the sensors if no serial could be fetched?
We could return a bool in setup and if not true then stop in line 65 with a meaningful error message.

Reply:
If it's a non critical error, we could `return None` for `unique_id` in this case. If it's an error that the user needs to fix or a critical unrecoverable error we should `return False` from component setup.

Reply:
so actually it is a critical error, as if there is no data fetched, there might be no connection to the charging station (e.g. ip address wrong in config file). I will fetch this and return False in the component setup.